README.md
lnkckr
======

**lnkckr** is a link checking library with command-line script `linkcheck`. It's intended to be used to check broken links and it can be interrupted at any time, then resumes from saved JSON file.

It supports Python 2.7 and Python 3.

**Contents**

[TOC]

Installation
------------

    pip install lnkckr

Usage
-----

    linkcheck [-c CHECKER] [-u|--update] [-f FILE] [-j JSON] [-s STATUS] [-x EXCLUDE_STATUS]

Arugment | Detail
--- | ---
`CHECKER` | what checker used to process `FILE`, see *Checkers* below.
`-u` or `--update` | update the JSON with input file. Wihout this, when `JSON` used, `FILE` is ignored.
`FILE` | the input filename or URL.
`JSON` | the filename of saved progress file. If the `FILE` is a filename, then `FILE` can be omitted, an filename is assigned automatically unless using different filename is desired.
`STATUS` | indicates re-check url with specific status.
`EXCLUDE_STATUS` | a CSV which lists what links with specified status will not be shown in report section. Default is `unchecked,200`.

Screenshot
----------

![Using lnkckr to check a Blogger XML Export file](https://lh5.googleusercontent.com/-JBBe-HVH_0M/URuaB0yeZHI/AAAAAAAAEc8/E6O7uL9gmJI/s800/lnkckr-blogger.png)

Checkers
--------

Checker | Description
--- | ---
`list` | input file is a list of links, one URL per line.
`html` | input file is a HTML file.
`blogger` | input file is a Blogger XML Export file.

### Blogger

The Blogger checker has the following environment variables:

Variable | Description | Default
--- | --- | ---
`LNKCKR_BLOGGER_TOPLIST_LIMIT` | the amount of entries listed in toplist. 0 means list all entries. | 10

Results
-------

Symbol | Description
--- | ---
`[---]` | the url hasn't been checked.
`[???]` | where the `???` is the HTTP status code.
`[###]` | means the fragment in the URL can't be found in the response body in format of `id="<fragment>"` or `name="<fragment>"`.
`[SCH]` | unsupported HTTP scheme.
`[SKP]` | the url is skipped.
`[RRR]` | the url results reaching maximal redirection limit.
`[XXX]` | unknown error.
`[000]` | timeout when trying to check the url.

Note
----

The output of lnkckr is similar to my two-year-old shell script `linkckr.sh`, I have also included it in the repository since lnkckr is kind of continuation of `linkckr.sh`, here is a [blog post about it](http://blog.yjl.im/2011/02/link-checker-bash-script-using-xmllint.html).

Related links
-------------

* [bea][]: Blogger Export Analyzer, not directly related, but also uses the Blogger XML Export file.
* [announcement][]: the blog post of lnkckr. 
* [b.py][]: Command-line posting script, it uses lnkckr to check links.

[bea]: https://bitbucket.org/livibetter/bea
[announcement]: http://blog.yjl.im/2013/02/checking-broken-links-with-blogger-xml.html
[b.py]: https://bitbucket.org/livibetter/b.py

License
-------

lnkckr is licensed under the MIT License, see `COPYING`.


CHANGES.md
CHANGES
=======

## Development

## Version 0.4.0 (2016-01-01T00:45:35Z)

* [base] really fix CPU hogging
* [base] add option `-U` (`--unverified-certificates`) for not verifying certifcates
* [base] output starte, ended, and elapsed times

## Version 0.3.0 (2015-07-31T23:18:21Z)

* [base] fix CPU hogging
* [base] fix fragment check, href=#frag can also be anchor
* [base] adjust URL quoting, only quote when there is `%` in URL
* [blogger] add `LNKCKR_BLOGGER_TOPLIST_LIMIT` to control the amount of listed entries
* [html] fix empty post content in Blogger XML causing `lxml.etree.XMLSyntaxError: line 1: b'Tag mn invalid'`
* [test] add empty content test

## Version 0.2.1 (2013-07-25T02:53:45Z)

* fix `all` option of `--status`.

## Version 0.2.0 (2013-06-11T22:53:15Z)

* add `-x` (`--exclude-status`) option for specifying excluded statuses from report section. (#2)

## Version 0.1.5 (2013-04-29T16:15:30Z)

* [base] fix query string getting quoted by mistake by e79523b
* [base] make HEADERS per instance not per class
* [test] introduce `X-MatchPath` to ensure path/querystring is correct

## Version 0.1.4 (2013-04-22T22:41:51Z)

 * [blogger] skip scheduled posts because of no post URLs
 * [base] fix encoding issue with URL with non-ISO-8859-1 characters in path

## Version 0.1.3 (2013-03-24T23:15:55Z)

 * add `--version` option
 * fix HTML unquoted attribute names for id and name matching

## Version 0.1.2 (2013-02-23T06:08:12Z)

 * [base] add `local_html` for checking local fragment
 * [html] use `local_html` to check local fragment
 * add `print_toplist` for a list of worse post/entry/etc
 * add HTML attribute `name` to search list of fragment and use regular expression to match
 * fix #1: url not checked when same url with fragment also being checked
 * url with fragment has return text/html type or it will be `[###]`
 * fragment now has to be valid ASCII or will not be found
 * add total numbers in summary

## Version 0.1.1 (2013-02-15T12:07:09Z)

 * add `--update` option to update JSON if input file is updated
 * urls with different fragments now only be checked once
 * ignore blank local fragment, i.e. url has only `#`
 * fix deadlock issue

## Version 0.1 (2013-02-13T13:18:32Z)

 * first version


tests/test_lnkckr_checkers.py
# Copyright (C) 2013 by Yu-Jie Lin
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.


import doctest

from lnkckr.checkers import modules


def load_tests(loader, tests, pattern):

  for module in modules:
    tests.addTests(doctest.DocTestSuite(module))
  return tests


tests/test_lnkckr_checkers_base.py
# -*- coding: utf-8 -*-
# Copyright (C) 2013 by Yu-Jie Lin
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.


try:
  from http.server import BaseHTTPRequestHandler, HTTPServer
except ImportError:
  from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer
try:
  from StringIO import StringIO
except ImportError:
  from io import StringIO
import json
import re
import unittest
from multiprocessing import Process, Value
from random import randint

from lnkckr.checkers.base import Checker

HOST = 'localhost'
# randomize the port for test server. why chose 32767 cap:
# http://en.wikipedia.org/wiki/Ephemeral_port
PORT = randint(8000, 32767)
H = 'http://%s:%s/' % (HOST, PORT)


class Handler(BaseHTTPRequestHandler):

  RE_CODE = re.compile(r'/(\d{3})')

  def do_GET(self):

    # counting the requests
    self.server.requests.value += 1
    redir = H
    m = self.RE_CODE.match(self.path)
    MatchPath = self.headers.get('X-MatchPath')
    if MatchPath and self.path != MatchPath:
      print('X-MatchPath: %s != %s' % (self.path, MatchPath))
      code = 404
    elif m:
      code = int(m.group(1))
    elif self.path == '/':
      code = 200
    elif self.path == '/loop':
      code = 302
      redir = '/loop'
    else:
      code = 404

    self.send_response(code)
    if code in (301, 302):
      self.send_header('Location', redir)
    else:
      self.send_header("Content-Type", "text/html")
    self.end_headers()

    if code == 200:
      data = self.headers.get('X-Echo')
      if data:
        self.wfile.write(data.encode('utf8'))

  def do_HEAD(self):

    self.do_GET()

  def log_message(self, *args):

    pass


def run_httpd(running, requests):

  httpd = HTTPServer((HOST, PORT), Handler)
  httpd.requests = requests
  httpd.requests.value = 0
  httpd.timeout = 0.01
  while running.value:
    httpd.handle_request()


class BaseCheckerTestCase(unittest.TestCase):

  @classmethod
  def setUpClass(self):

    self.httpd_running = Value('b', 1)
    self.httpd_requests = Value('L', 1)
    args = (self.httpd_running, self.httpd_requests)
    self.httpd_process = Process(target=run_httpd, args=args)
    self.httpd_process.start()

  @classmethod
  def tearDownClass(self):

    self.httpd_running.value = 0
    self.httpd_process.join()

  def setUp(self):

    self.checker = Checker()
    self.checker.do_update = lambda url, link: None

  def tearDown(self):

    self.checker = None

  def test_init(self):

    self.assertEqual(self.checker.links, {})

  # =====

  def test_load_json(self):

    checker = self.checker

    data = {'foo': 'bar'}
    links = {'http://example.com': {'status': None, 'data': 'foobar'}}
    jsondata = {'data': data, 'links': links}

    checker.load_json(StringIO(json.dumps(jsondata)))

    self.assertEqual(checker.data, data)
    self.assertEqual(checker.links, links)

  def test_save_json(self):

    checker = self.checker
    dest = StringIO()

    data = {'foo': 'bar'}
    links = {'http://example.com': {'status': None, 'data': 'foobar'}}
    jsondata = {'data': data, 'links': links}

    checker.data = data.copy()
    checker.add_link('http://example.com', {'data': 'foobar'})
    checker.save_json(dest)

    self.assertEqual(dest.getvalue(), json.dumps(jsondata))

  # =====

  def test_add_link_local_blank_fragment(self):

    checker = self.checker

    checker.add_link('#')
    self.assertEqual(checker.links, {})

  # =====

  def test_update_links(self):

    checker = self.checker
    checker.add_link('http://example.com', {'data': 'foobar'})
    checker.update_links({})
    self.assertEqual(checker.links, {})

    new_links = {'http://example.com/': {'status': None, 'data': 'blah'}}
    checker.update_links(new_links)
    self.assertEqual(checker.links, new_links)

    new_links = {'http://example.com/': {'status': 123, 'data': 'duh'}}
    expect = {'http://example.com/': {'status': None, 'data': 'duh'}}
    checker.update_links(new_links)
    self.assertEqual(checker.links, expect)

    checker.update_links(new_links, update_status=True)
    self.assertEqual(checker.links, new_links)

  # =====

  def test_check(self):

    checker = self.checker

    checker.add_link(H + '200')
    checker.check()
    expect = {H + '200': {'status': '200', 'redirection': None}}
    self.assertEqual(checker.links, expect)

    checker.add_link(H + '200')
    checker.add_link(H + '301')
    checker.check(lambda item: item[0].endswith('301'))
    expect = {
      H + '200': {
        'status': None,
      },
      H + '301': {
        'status': '200',
        'redirection': H,
      },
    }
    self.assertEqual(checker.links, expect)

  def test_check_max_redirection(self):

    checker = self.checker

    checker.add_link(H + 'loop')
    checker.check()
    expect = {H + 'loop': {'status': 'RRR', 'redirection': H + 'loop'}}
    self.assertEqual(checker.links, expect)

  def test_check_non_iso8859_1(self):

    checker = self.checker

    checker.add_link(H + '200/áéíóú')
    checker.HEADERS['X-MatchPath'] = '/200/%C3%A1%C3%A9%C3%AD%C3%B3%C3%BA'
    checker.check()
    expect = {H + '200/áéíóú': {'status': '200', 'redirection': None}}
    self.assertEqual(checker.links, expect)

  def test_check_query(self):

    checker = self.checker

    checker.add_link(H + '200?id=1234')
    checker.HEADERS['X-MatchPath'] = '/200?id=1234'
    checker.check()
    expect = {H + '200?id=1234': {'status': '200', 'redirection': None}}
    self.assertEqual(checker.links, expect)

  # -----

  def test_check_fragment(self):

    checker = self.checker

    checker.add_link(H + '200#foobar')
    checker.HEADERS['X-Echo'] = '<h1 id="foobar">blah</h1>'
    checker.check()
    expect = {H + '200#foobar': {'status': '200', 'redirection': None}}
    self.assertEqual(checker.links, expect)

  def test_check_fragment_missing(self):

    checker = self.checker

    checker.add_link(H + '200#foobar')
    checker.HEADERS['X-Echo'] = '<h1>blah</h1>'
    checker.check()
    expect = {H + '200#foobar': {'status': '###', 'redirection': None}}
    self.assertEqual(checker.links, expect)

  def test_check_fragments(self):
    """Test same url with different fragments and only make HTTP request
    once.
    """
    checker = self.checker

    req_count = self.httpd_requests.value
    checker.add_link(H + '200#foobar')
    checker.add_link(H + '200#foobar1')
    checker.HEADERS['X-Echo'] = '<h1 id="foobar">blah</h1>'
    checker.check()

    expect = {
      H + '200#foobar': {'status': '200', 'redirection': None},
      H + '200#foobar1': {'status': '###', 'redirection': None},
    }
    self.assertEqual(checker.links, expect)
    self.assertEqual(self.httpd_requests.value, req_count + 1)

  def test_check_fragments_1(self):
    """Test same url with and without fragments and only make HTTP request
    once.
    """
    checker = self.checker

    req_count = self.httpd_requests.value
    checker.add_link(H + '200')
    checker.add_link(H + '200#foobar')
    checker.HEADERS['X-Echo'] = '<h1 id="foobar">blah</h1>'
    checker.check()

    expect = {
      H + '200': {'status': '200', 'redirection': None},
      H + '200#foobar': {'status': '200', 'redirection': None},
    }
    self.assertEqual(checker.links, expect)
    self.assertEqual(self.httpd_requests.value, req_count + 1)

  def test_check_local_fragments(self):
    """Test local fragments and if make no HTTP requests"""
    checker = self.checker

    req_count = self.httpd_requests.value
    checker.add_link('#foo1')
    checker.add_link('#foo2')
    checker.check()

    expect = {
      '#foo1': {'status': 'SCH', 'redirection': None},
      '#foo2': {'status': 'SCH', 'redirection': None},
    }
    self.assertEqual(checker.links, expect)
    self.assertEqual(self.httpd_requests.value, req_count)


tests/test_lnkckr_checkers_html.py
# Copyright (c) 2013, 2015 Yu-Jie Lin
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.


try:
  from StringIO import StringIO
except ImportError:
  from io import StringIO

from lnkckr.checkers.html import Checker
from test_lnkckr_checkers_base import H, BaseCheckerTestCase


class HTMLCheckerTestCase(BaseCheckerTestCase):

  def setUp(self):

    self.checker = Checker()
    self.checker.do_update = lambda url, link: None

  # =====

  def test_load(self):

    checker = self.checker

    codes = (H + '200', H + '403', H + '404')
    html = '\n'.join('<a href="%s">text</a>' % href for href in codes)
    src = StringIO(html)
    checker.load(src)
    expect = {
      H + '200': {'status': None},
      H + '403': {'status': None},
      H + '404': {'status': None},
    }
    self.assertEqual(checker.links, expect)

    checker.check()
    expect = {
      H + '200': {'status': '200', 'redirection': None},
      H + '403': {'status': '403', 'redirection': None},
      H + '404': {'status': '404', 'redirection': None},
    }
    self.assertEqual(checker.links, expect)

  # =====

  def test_local_html(self):

    checker = self.checker

    html = '<a href="#foo1">blah</a><span id="foo1">blah</span>'
    html += '<a href="#foo2">blah</a>'
    checker.process(StringIO(html))

    checker.check()
    expect = {
      '#foo1': {'status': '200', 'redirection': None},
      '#foo2': {'status': '200', 'redirection': None},
    }
    self.assertEqual(checker.links, expect)

  # =====

  def test_empty_html(self):

    checker = self.checker

    html = ''
    checker.process(StringIO(html))

    checker.check()
    expect = {}
    self.assertEqual(checker.links, expect)

    html = None
    checker.process(StringIO(html))

    checker.check()
    expect = {}
    self.assertEqual(checker.links, expect)


lnkckr/checkers/base.py
# Copyright (c) 2013, 2015 Yu-Jie Lin
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.


from __future__ import print_function

import json
import re
import socket
import ssl
import time
import traceback
from itertools import groupby, product
from multiprocessing import Manager, Process, Queue, Value
from os import path

import lnkckr

try:
  from http.client import HTTPConnection, HTTPSConnection
except ImportError:
  from httplib import HTTPConnection, HTTPSConnection
try:
  from queue import Empty, Full
except ImportError:
  from Queue import Empty, Full
try:
  from urllib.parse import quote, urldefrag, urljoin, urlparse
except ImportError:
  from urllib import quote
  from urlparse import urldefrag, urljoin, urlparse
try:
  from urllib.request import urlopen
except ImportError:
  from urllib import urlopen


class Checker():
  """Base checker"""
  ID = None

  MAX_WORKERS = 10
  QUEUE_SIZE = 20
  SAVE_INT = 100

  def __init__(self, **cfg):

    self.data = {}
    self.links = {}
    self.json_filename = None

    self.exclude_status = cfg.get('exclude_status', (None, 200))
    self.unverified_certificates = cfg.get('unverified_certificates', False)

    # User-Agent for some website like Wikipedia. Without it, most of requests
    # result in 403.
    self.HEADERS = {
      'User-Agent': '%s/%s' % (lnkckr.__name__, lnkckr.__version__)
    }

  def load(self, src=None, jsonsrc=None, do_update=False, update_status=False):
    """Load links from a file

    src can be a URL (starting with http), file-like object, or a filename.
    jsonsrc can be a file-like or a filename.

    "src.json" may be saved if src is a file and jsonsrc isn't a file.

    do_update will update the JSON.

    The file format depends how self.process() is implemented."""
    if jsonsrc:
      self.json_filename = self.load_json(jsonsrc)
      if not do_update:
        return
      json_links = self.links

    if not src:
      return

    f = None
    try:
      if hasattr(src, 'startswith') and (src.startswith('http://') or
                                         src.startswith('https://')):
        f = urlopen(src)
      elif hasattr(src, 'read'):
        f = src
      else:
        if not self.json_filename:
          self.json_filename = src + '.json'
          if not do_update and path.exists(self.json_filename):
            self.load_json(self.json_filename)
            return
        f = open(src, 'r')

      self.process(f)
    finally:
      if f:
        f.close()

    if jsonsrc and do_update:
      file_links = self.links
      self.links = json_links
      self.update_links(file_links, update_status)

  def process(self, data):
    """Process the data from load()"""
    pass

  def load_json(self, src):
    """Load data and links from a JSON file

    src can be a filename or file-like object.

    returns src only if it's file and load successfully.
    """
    filename = None
    if hasattr(src, 'read'):
      jsondata = json.load(src)
      src.close()
    else:
      with open(src, 'r') as f:
        jsondata = json.load(f)
        filename = src

    # version <= 0.1.1
    if 'data' not in jsondata or 'links' not in jsondata:
      self.links = jsondata
    else:
      self.data = jsondata['data']
      self.links = jsondata['links']

    return filename

  def save_json(self, dest):
    """Save data and links to a JSON file

    dest can be a filename or a file-like object, will not be closed by
    save_json if it's a file-like object.
    """
    jsondata = {'data': self.data, 'links': self.links}
    if hasattr(dest, 'write'):
      json.dump(jsondata, dest)
      return

    with open(dest, 'w') as f:
      json.dump(jsondata, f)

  def add_link(self, url, data=None):
    """Add a link

    If the link already exists, it will be replaced.

    >>> c = Checker()
    >>> c.add_link('http://example.com')
    >>> c.links
    {'http://example.com': {'status': None}}
    >>> c.add_link('http://example.com', {'data': 'foobar'})
    >>> c.links == {'http://example.com': {'status': None, 'data': 'foobar'}}
    True
    """
    # skip blank fragment
    if url == '#':
      return
    d = {'status': None}
    d.update(data or {})
    self.links.update({url: d})

  def update_links(self, links, update_status=False):
    """Update self.links with links"""
    urls = set(self.links.keys())
    nurls = set(links.keys())

    # remove removed links
    for url in urls - nurls:
      del self.links[url]
    # add new links
    for url in nurls - urls:
      self.links[url] = links[url]
    # update links
    for url in urls & nurls:
      self.update_link(url, links[url], update_status)

  def update_link(self, url, new_link, update_status=False):
    """Update self.links[url] with new_link

    >>> c = Checker()
    >>> url = 'http://example.com/'
    >>> c.add_link(url)
    >>> new_link = {'status': 123, 'foobar': 'blah'}
    >>> c.update_link(url, new_link)
    >>> c.links[url] == {'status': None, 'foobar': 'blah'}
    True
    >>> new_link['foobar'] = 'duh'
    >>> c.update_link(url, new_link, update_status=True)
    >>> c.links[url] == {'status': 123, 'foobar': 'duh'}
    True
    """
    link = self.links[url]
    status = link['status']
    link.update(new_link)
    if not update_status:
      link['status'] = status

  # =====

  # http://www.w3.org/TR/html-markup/syntax.html#syntax-attributes
  RE_ID_NAME = re.compile(r'\b(?:id|name) *= *'
                          r'(?:(?P<q>[\'"])(.+?)(?P=q)|([^"\'=><`]+))', re.I)
  RE_HREF = re.compile(r'\bhref *= *'
                       r'(?:(?P<q>[\'"])#(.+?)(?P=q)|#([^"\'=><`]+))', re.I)

  def _check_url_frag(self, data):
    """search for fragment in HTML attributes id or name look-likes

    return '200' if found else '###'

    >>> c = Checker()
    >>> f = c._check_url_frag
    >>> f(('', 'blah'))
    '200'
    >>> f(('foo', 'id=foo'))
    '200'
    >>> f(('foo', 'id="foo"'))
    '200'
    >>> f(('foo', 'id="nofoo"'))
    '###'
    >>> f(('bar', 'name=bar'))
    '200'
    >>> f(('bar', 'name="bar"'))
    '200'
    >>> f(('foo', 'href="#foo"'))
    '200'
    >>> f(('foo', 'href=foo'))
    '###'
    """
    frag, body = data
    if frag == '':
      return '200'
    for m in self.RE_ID_NAME.finditer(body):
      if frag == m.group(2) or m.group(3):
        return '200'
    for m in self.RE_HREF.finditer(body):
      if frag == m.group(2) or m.group(3):
        return '200'

    return '###'

  def check_url(self, url, frags=None, local_html=None):
    """Check a url

    Returns (final status code(s) in tuple of string, final redirect url)

    >>> c = Checker()
    >>> c.check_url('http://example.com') # doctest: +SKIP
    (('200',), 'http://www.iana.org/domains/example')
    """
    MAX_REDIRS = 10
    redirs = 0
    method = 'HEAD'
    start_url = url
    status = ''
    statuses = []

    while not status:
      if url.startswith('//'):
        url = 'http:' + url
      url_comp = urlparse(url)
      if url_comp.scheme == 'http':
        conn = HTTPConnection(url_comp.netloc)
      elif url_comp.scheme == 'https':
        if self.unverified_certificates:
          context = ssl._create_unverified_context()
        else:
          context = ssl.create_default_context()
        conn = HTTPSConnection(url_comp.netloc, context=context)
      else:
        if not url and frags and local_html:
          pairs = product(frags, (local_html,))
          statuses = tuple(map(self._check_url_frag, pairs))
          break
        status = 'SCH'
        if url_comp.scheme in ('about', 'javascript'):
          status = 'SKP'
        break

      try:
        # attempt to decide whether lnkckr should quote the URL.
        # if there is '%' in URL, then don't quote it.
        p = url_comp.path if '%' in url_comp.path else quote(url_comp.path)
        if url_comp.query:
          p += '?' + url_comp.query
        if frags:
          method = 'GET'
        conn.request(method, p, headers=self.HEADERS)
        resp = conn.getresponse()
        if resp.status == 200 and frags != ('',):
          # any non text/html result ### as such fragment isn't found
          if not resp.getheader('Content-Type', '').startswith('text/html'):
            status = '###'
            break
          # assume characters in fragment are valid ASCII
          rbody = resp.read().decode('ascii', 'ignore')
          pairs = product(frags, (rbody,))
          statuses = tuple(map(self._check_url_frag, pairs))
          break
        elif 300 <= resp.status < 400:
          if redirs >= MAX_REDIRS:
            status = 'RRR'
          else:
            redirs += 1
            url = urljoin(url, resp.getheader('location'))
            method = 'HEAD'
        elif resp.status == 405 and method == 'HEAD':
          method = 'GET'
        else:
          status = str(resp.status)
        conn.close()
      except socket.error:
        status = '000'
      except Exception:
        traceback.print_exc()
        status = 'XXX'

    if start_url == url and not redirs:
      url = None
    if not statuses:
      statuses = (status,) * len(frags)
    return statuses, url

  def check_worker(self, q, r, running, data):

    local_html = data.get('local_html', '')
    try:
      while not q.empty() or running.value:
        try:
          url, frags = q.get(timeout=0.01)
          statuses, rurl = self.check_url(url, frags, local_html=local_html)
          for frag, status in zip(frags, statuses):
            data = (url + '#' + frag if frag else url, (status, rurl))
            while data:
              try:
                r.put(data, False, 0.01)
                data = None
              except Full:
                pass
          continue
        except Empty:
          pass
    except KeyboardInterrupt:
      pass

  def check_update_links(self, r):

    while not r.empty():
      url, data = r.get(block=False)
      link = self.links[url]
      status, final_url = data
      link['status'] = status
      link['redirection'] = final_url
      self.do_update(url, link)
      return True

  def _check_groupby(self, urls):
    """Generator to group by parts of url without the fragment

    The fragments will collected as a tuple.

    >>> c = Checker()
    >>> urls = ['http://example.com']
    >>> list(c._check_groupby(urls))
    [('http://example.com', ('',))]
    >>> urls = [
    ...   'http://example.com',
    ...   'http://example.com/foo#bar1',
    ...   'http://example.com/foo#bar2',
    ...   'http://example.com/foobar',
    ...   'http://example.com/foobar#blah',
    ... ]
    >>> list(c._check_groupby(urls)) # doctest: +NORMALIZE_WHITESPACE
    [('http://example.com', ('',)),
     ('http://example.com/foo', ('bar1', 'bar2')),
     ('http://example.com/foobar', ('', 'blah'))]
    >>> urls = ['#foo1', '#foo2']
    >>> list(c._check_groupby(urls))
    [('', ('foo1', 'foo2'))]
    """
    key = lambda item: item[0]
    for url, g in groupby(sorted(map(urldefrag, urls), key=key), key):
      yield url, tuple(item[1] for item in g)

  def check(self, f=None):
    """Check links

    f is a function for passing to filter function to filter links with
    argument (url, link).
    """
    default_timeout = socket.getdefaulttimeout()
    q = Queue(self.QUEUE_SIZE)
    r = Queue(self.QUEUE_SIZE)
    running = Value('b', 1)
    workers = []
    manager = Manager()
    datadict = manager.dict(self.data)
    for i in range(self.MAX_WORKERS):
      worker = Process(name='checker #%d' % i,
                       target=self.check_worker,
                       args=(q, r, running, datadict))
      worker.start()
      workers.append(worker)

    total = 0
    count = 0
    try:
      if f is None:
        f = lambda item: item[1]['status'] is None
      urls = (item[0] for item in filter(f, self.links.items()))
      gurls = self._check_groupby(urls)
      for idx, item in enumerate(gurls, start=1):
        total += 1 if not item[1] else len(item[1])
        while item:
          try:
            q.put(item, False, 0.01)
            item = None
          except Full:
            count += 1 if self.check_update_links(r) else 0
          time.sleep(0.001)
        count += 1 if self.check_update_links(r) else 0
        if count % self.SAVE_INT == 0:
          self.do_save()
      while count < total:
        count += 1 if self.check_update_links(r) else 0
        if count % self.SAVE_INT == 0:
          self.do_save()
        time.sleep(0.001)
    except KeyboardInterrupt:
      pass

    running.value = 0
    for worker in workers:
      worker.join()
    self.check_update_links(r)
    self.do_save()
    socket.setdefaulttimeout(default_timeout)

  def do_update(self, url, link):
    """Call by update_links when a link is updated"""
    self.format_status(url, link)

  def do_save(self):
    """Call by check for saving JSON when necessary"""
    if self.json_filename:
      self.save_json(self.json_filename)

  ###########################
  # report output functions #
  ###########################

  def color_status(self, status):

    if status == '000':
      return '\033[1;36m[%s]\033[0m' % status
    if status == '200':
      return '\033[1;32m[%s]\033[0m' % status
    if status in ('SKP', 'SCH'):
      return '\033[1;33m[%s]\033[0m' % status
    else:
      return '\033[1;31m[%s]\033[0m' % status

  def format_status(self, url, link):

    status = link['status']
    redir = link['redirection']
    print('%s %s' % (self.color_status(status), url), end='')
    if redir:
      print(' \033[1;33m\n   ->\033[0m %s' % redir, end='')
    print()

  def num_len(self, n):
    """Return length of n in string including thousands separator

    >>> c = Checker()
    >>> c.num_len(123)
    3
    >>> c.num_len(1234)
    5
    """
    return len('{:,d}'.format(n))

  def print_heading(self, text):

    bar = '{0:{f}^{l}s}'.format('', f='=', l=len(text) + 4)
    print('{0}\n= {1} =\n{0}\n'.format(bar, text))

  def print_all(self):

    self.print_report()
    self.print_summary()

  # report
  #########

  def print_report(self):

    self.print_heading('report')

    links = self.links

    unchecked = 0
    key = lambda item: item[1]['status'] or '---'
    for status, g in groupby(sorted(links.items(), key=key), key=key):
      if status == '---':
        unchecked = len(list(g))
        continue
      for url, link in sorted(g):
        self.print_report_link(url, link)
    if unchecked:
      print('*** checking process is not finished, '
            '%d links have not been checked. ***' % unchecked)
    print()

  def print_report_link(self, url, link):

    if link['status'] in self.exclude_status:
      return
    self.format_status(url, link)
    self.print_report_link_data(url, link)

  def print_report_link_data(self, url, link):

    pass

  # summary
  ##########

  def print_summary(self):

    self.print_heading('summary')

    key = lambda link: link['status'] or '---'
    data = []
    for status, g in groupby(sorted(self.links.values(), key=key), key=key):
      data.append((status, list(g)))
    self.print_summary_status(data)
    self.print_summary_footer()

  def print_summary_status(self, data):

    data2 = []
    for status, links in data:
      nlinks = sum(1 for l in links)
      data2.append((status, nlinks))

    l = max(self.num_len(item[1]) for item in data2) if data2 else 0

    for status, nlinks in data2:
      print('{} {:{l},d} links'.format(self.color_status(status), nlinks, l=l))

  def print_summary_footer(self):

    print()
    print('TOTAL {:,} links'.format(len(self.links)))
    print()
